lib/lz4.c

```diff
@@ -602,8 +640,19 @@ LZ4_FORCE_INLINE int LZ4_compress_generic(
                 if (unlikely(forwardIp > mflimit)) goto _last_literals;
 
                 match = LZ4_getPositionOnHash(h, cctx->hashTable, tableType, base);
-                if (dict==usingExtDict) {
+                if (dictDirective == usingExtDictCtx) {
                     if (match < (const BYTE*)source) {
+                        /* there was no match, try the dictionary */
+                        /* TODO: use precalc-ed hash? */
```

Comment:
That's an interesting optimization.
Actually, there are 2 scenarios : 
- use same hashing formula, hence re-use hash value
- use this opportunity to have _different_ hashing formula. For example, if dictionary uses a `byU16` table, while streaming data use `byU32` instead, dictionary hash values use one more bit. It's also possible deliberately target different `minmatch` sizes (LZ4 already automatically selects 4 bytes or 5 bytes depending on scenario).

Reply:
The issue with this is that if we're going to copy the dict table into the compression ctx, we need it to be `tableType`-compatible (i.e., `byU32`). If we're going to keep them separate, it would be better if it were `byU16`. We don't know which scenario we're in though until we know the size of the blob we're compressing, which is not known at dict load time. So I'm not sure there's a good solution.

Reply:
Is there still a size threshold after which it's preferable to copy the table rather than using it in-place ?

Reply:
Right now, I'm being pretty conservative and copying the table over if the blob to compress is larger than 2KB. I don't think there's a clear line, but it's somewhere between 1KB and 32KB.

Reply:
We will have to measure again this threshold once optimisation pass is completed.